#+TITLE: bcbioRNASeq
#+STARTUP: content
* Development
** TODO Need to handle edge case of kallisto usage with GENCODE.
    Need to split based on the pipe and take the first instance for the transcript ID.
    https://support.bioconductor.org/p/9149475/
** TODO Need to improve code coverage / error messages against `ensemblRelease` mismatch.
    https://github.com/hbc/bcbioRNASeq/issues/190
** TODO R Markdown sidebar now is messed up in render output.
    Consider reworking AcidMarkdown defaults to address this.
** TODO Consider rounding raw counts to integer.
    This better matches conventions used in internal bcbio-nextgen R script.
    Non-integer counts tend to confuse people.
    Only want to implement this at gene level.
** TODO Inform the user in the QC template which aligner was used for QC (e.g. STAR, HISAT2).
** TODO Look into tximeta integration.
** TODO Consider adding an alert message that fast mode doesn't support functional analysis.
    Recommend that organism, genomeBuild, ensemblRelease are defined.
** TODO Improve error message when user passes in incorrect genome.
    e.g. Ensembl 105 when using bcbio default hg38 genome.
    See logs from 2022-04-05 for debugging.
** TODO Consider reexporting defunct flatFiles here so we don't break bcbio 1.2.8.
** TODO In bcbioRNASeq class checks, consider allowing NULL tx2gene when level = "transcripts".
** TODO In internal `.tximport` function, need to dynamically set `ignoreTxVersion`?
** TODO In internal `.tximport` function, allow NULL tx2gene input when level = "transcripts".
** TODO Need to improve code coverage of kallisto and sailfish import.
** TODO Supporting loading of multiple bcbio runs into a single object?
    https://github.com/hbc/bcbioRNASeq/issues/141
** TODO Transcript level analysis with sleuth.
    This is supported but we need to improve documentation on how to hand off to sleuth from the bcbioRNASeq object.
** TODO DTU with DRIMSeq and DEXSeq.
    Need to document how to do handoff from the bcbioRNASeq object for this.
** TODO Likelihood ratio test support.
    Need to improve documentation and/or add an R Markdown for this.
    https://github.com/hbc/bcbioRNASeq/issues/145
** TODO Require valid names in all slots.
** TODO Consider renaming "normalized" assay to "sizeFactor" or simply "sf".
** TODO plotMeanSD edgeR RLE is messed up for new LnCap data.
** TODO Consider slotting aligned counts into object automatically in updateObject call.
** TODO `plotGeneSaturation()`: Consider adding y-intercept 0.
** TODO bcbioRNASeq: Improve error message when user doesn't point to `final/` directory.
    Or make this more flexible and detect `final/` automatically.
** TODO `plotMappingRate()`: Plot exonic, intronic, intergenic rates.
    Show these similarly to qualimap output in MultiQC report.
** TODO `plotCountsPerGene()`: Don't hide the sample legend for `sampleName` interesting groups.
** TODO Add RSEM support for tximport? Is this supported in bcbio?
** TODO Work on additional multiQC-like plots.
    `plotInsertSize()`, `plotDuplicatedReads()`, `plotProperlyPairedReads()`,
    `plotReadsAligned()`, `plotGCContent()`, `plotXYCounts()`.
** TODO `updateObject()`: Add update method for run-length encoding in rowRanges.
* F1000 manuscript revisions
** TODO Switch paper to use `deg()` instead of `significants()`.
* Miscellaneous
** TODO Consider running BFG on the repo to remove old binary commits.
