---
## Alternatively, consider using the AcidGSEA package for fast preranked GSEA.
##
## Expecting that input DESeq2 analysis uses Ensembl identifiers.
##
## - KEGG supported organisms:
##   http://www.genome.jp/kegg/catalog/org_list.html
## - OrgDb supported organisms:
##   http://bioconductor.org/packages/release/BiocViews.html#___OrgDb
##
## Refer to `rmarkdown::render()` for details on how to generate an HTML report
## from this template inside of an R script. In particular, the `params`
## argument is useful for passing in variables.
##
## Updated 2022-05-26.
params:
  title: "Functional analysis with clusterProfiler"

  ## DESeqAnalysis object file.
  ## Refer to DESeqAnalysis package for details.
  ## This is *not* the DESeqDataSet object.
  deseq_file: !r file.path("data", "deseq.rds")

  ## Results (i.e. contrast) name or position.
  ## Refer to `resultsNames()` for details.
  results_name: "treatment_vs_control"

  ## Full Latin organism name (e.g. "Homo sapiens"). If left `NULL`, will
  ## attempt to detect from DESeqAnalysis object.
  organism: NULL

  ## Supported GO classes:
  ## - "BP": Biological Process
  ## - "MF": Molecular Function
  ## - "CC": Cellular Component
  go_class: "BP"

  ## Directory where R data files will be saved.
  data_dir: !r file.path("data")

  ## Directory where results, including CSV files, will be saved.
  results_dir: !r file.path("results", Sys.Date(), "functional-analysis")

  ## Pathview KEGG plots tend to be slow to render.
  pathview: FALSE

title: "`r params[['title']]`"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"

## This file is generated by the `prepareTemplate()` step in setup chunk.
bibliography: "bibliography.bib"
---

```{r setup, cache=FALSE, message=FALSE}
## nolint start
suppressPackageStartupMessages({
    library(goalie)
    library(stringi)
    library(basejump)
    library(bcbioRNASeq)
    library(DESeqAnalysis)
    library(AcidGSEA)
})
prepareTemplate()
source("_setup.R")
## May encounter warning from ggrepel package otherwise.
options("ggrepel.max.overlaps" = Inf)
## nolint end
keggPlotsDir <- file.path(params[["results_dir"]], "kegg-plots")
invisible(lapply(
    X = c(
        params[["data_dir"]],
        params[["results_dir"]],
        keggPlotsDir
    ),
    FUN = initDir
))
evalPathview <- params[["pathview"]]
```

```{r header, child="_header.Rmd"}
```

# Prepare variables for enrichment analysis

First, we need to load the `DESeqAnalysis` object, which has been generated
from our `bcbioRNASeq` object during the differential expression step.

```{r load-deseqanalysis}
deseq <- import(params[["deseq_file"]])
assert(
    is(deseq, "DESeqAnalysis"),
    validObject(deseq)
)
print(deseq)
```

We need to attach organism-specific packages from Bioconductor.

```{r bioc-packages, message=FALSE}
## Detect the organism from the DESeqDataSet object, if necessary.
organism <- params[["organism"]]
if (is.null(organism)) {
    organism <- organism(as(deseq, "DESeqDataSet"))
}
## Get OrgDb package from organism name (e.g. "org.Mm.eg.db").
match <- stringi::stri_match_first_regex(
    str = organism,
    pattern = "^([A-Z])[a-z]+\\s([a-z])[a-z]+$"
)
orgDb <- paste(
    "org",
    paste(match[, 2L:3L], collapse = ""),
    "eg",
    "db",
    sep = "."
)
rm(match)
## Use `BiocManager::install()` to install missing packages, if necessary.
packages <- c(
    "DOSE",
    "R.utils",
    "clusterProfiler",
    "enrichplot",
    "ggnewscale",  # For `cnetplot()`.
    "pathview",
    "withr",
    orgDb
)
assert(allAreInstalled(packages))
## Avoid logical coercion error in DOSE.
## nolint start
Sys.setenv(
    "_R_CHECK_LENGTH_1_CONDITION_" = "false",
    "_R_CHECK_LENGTH_1_LOGIC2_" = "false"
)
## nolint end
```

Match the KEGG organism code (e.g. "hsa" for Homo sapiens).

```{r check-db-codes}
keggCode <-
    organism |>
    clusterProfiler::search_kegg_organism("scientific_name") |>
    getElement("kegg_code")
assert(
    hasLength(keggCode, n = 1L),
    isMatchingRegex(pattern = "^[a-z]{3}$", x = keggCode),
    isSubset(params[["go_class"]], c("BP", "CC", "MF"))
)
```

We need to extract the corresponding `DESeqDataSet` and `DESeqResults` objects.

```{r get-deseqdataset}
dds <- as.DESeqDataSet(deseq)
assert(
    is(dds, "DESeqDataSet"),
    validObject(dds)
)
print(dds)
```

```{r get-deseqresults}
## Refer to `DESeqAnalysis::resultsNames()` for available options.
res <- DESeqAnalysis::results(object = deseq, i = params[["results_name"]])
assert(
    is(res, "DESeqResults"),
    validObject(res)
)
print(res)
```

Now we can prepare the gene list vectors required for enrichment analysis.

```{r results}
## Not allowing the user to apply post-hoc LFC filtering in params.
alphaThreshold <- alphaThreshold(res)
lfcThreshold <- lfcThreshold(res)
## Subset NA adjusted P values.
res <- res[!is.na(res[["padj"]]), , drop = FALSE]
res <- res[order(res[["padj"]]), , drop = FALSE]
## Match the DESeqDataSet.
dds <- dds[rownames(res), , drop = FALSE]
## Set our background genes vector (aka universe).
allGenes <- rownames(res)
## > help(topic = "deg", package = "DESeqAnalysis")
sigGenes <- deg(res, direction = "both")
sigRes <- res[sigGenes, , drop = FALSE]
## LFC ranked list vector.
lfcVec <- sigRes[["log2FoldChange"]]
names(lfcVec) <- rownames(sigRes)
lfcVec <- sort(lfcVec, decreasing = TRUE)
summary(lfcVec)
```

# GO enrichment analysis

[Gene Ontology (GO)][GO] term enrichment is a technique for interpreting sets
of genes making use of the [Gene Ontology][GO] system of classification, in
which genes are assigned to a set of predefined bins depending on their
functional characteristics.

```{r enrich-go}
enrichGo <- clusterProfiler::enrichGO(
    gene = sigGenes,
    OrgDb = orgDb,
    keyType = "ENSEMBL",
    ont = params[["go_class"]],
    universe = allGenes,
    qvalueCutoff = 0.05,
    readable = TRUE
)
enrichGoTbl <-
    enrichGo |>
    slot("result") |>
    as.data.frame() |>
    camelCase()
saveData(
    enrichGo,
    enrichGoTbl,
    dir = params[["data_dir"]]
)
export(
    object = enrichGoTbl,
    file = file.path(
        params[["results_dir"]],
        paste0(
            paste(
                "go",
                tolower(params[["go_class"]]),
                "clusterprofiler",
                "padj",
                alphaThreshold,
                "lfc",
                lfcThreshold,
                sep = "_"
            ),
            ".csv.gz"
        )
    )
)
print(enrichGoTbl)
```

## Dot plot

```{r dotplot}
try({
    suppressWarnings({
        enrichplot::dotplot(enrichGo, showCategory = 25L)
    })
})
```

## GO terms map

```{r emapplot}
try({
    suppressWarnings({
        enrichplot::emapplot(
            x = enrichplot::pairwise_termsim(enrichGo),
            showCategory = 25L
        )
    })
})
```

## Gene map

In order to consider the potentially biological complexities in which a gene
may belong to multiple annotation categories, and provide information of
numeric changes if available.

Here we are plotting genes colored by LFC for top significant [GO][] terms.

```{r cnetplot}
try({
    suppressWarnings({
        enrichplot::cnetplot(enrichGo, foldChange = lfcVec)
    })
})
```

## GO GSEA analysis

A common approach in analyzing gene expression profiles was identifying
differential expressed genes that are deemed interesting. The enrichment
analysis we demonstrated previously were based on these differentially
expressed genes. This approach will find genes where the difference is large,
but it will not detect a situation where the difference is small, but evidenced
in coordinated way in a set of related genes. [Gene Set Enrichment Analysis
(GSEA)][GSEA] directly addresses this limitation. All genes can be used in
[GSEA][]; [GSEA][] aggregates the per gene statistics across genes within a
gene set, therefore making it possible to detect situations where all genes in
a predefined set change in a small but coordinated way. Since it is likely that
many relevant phenotypic differences are manifested by small but consistent
changes in a set of genes.

```{r ranked-list}
rankedListEnsembl <- AcidGSEA::RankedList(
    object = deseq,
    value = "stat",
    keyType = "ensemblId"
)
rankedListEnsembl <- rankedListEnsembl[[params[["results_name"]]]]
assert(
    is.numeric(rankedListEnsembl),
    hasNames(rankedListEnsembl)
)
## Entrez identifiers must be defined in the "entrezId" column of DESeqDataSet
## rowData, otherwise this step will return NULL. In this case, we'll skip the
## enrichment steps that require Entrez identifiers (e.g. KEGG, see below).
rankedListEntrez <-
    tryCatch(
        expr = {
            x <- AcidGSEA::RankedList(
                object = deseq,
                value = "stat",
                keyType = "entrezId"
            )
            x <- x[[params[["results_name"]]]]
            x
        },
        error = function(e) NULL
    )
evalEntrez <- is.numeric(rankedListEntrez)
if (isFALSE(evalEntrez)) {
    evalPathview <- FALSE
}
```

```{r gse-go}
## Now we're ready to run GSEA.
## > help("gseGO", "clusterProfiler")
gseaGo <- clusterProfiler::gseGO(
    geneList = rankedListEnsembl,
    ont = params[["go_class"]],
    OrgDb = get(orgDb, envir = asNamespace(orgDb)),
    keyType = "ENSEMBL",
    pvalueCutoff = 0.05
)
gseaGoTbl <-
    gseaGo |>
    slot("result") |>
    as.data.frame() |>
    camelCase()
saveData(
    gseaGo, gseaGoTbl,
    dir = params[["data_dir"]]
)
export(
    object = gseaGoTbl,
    file = file.path(
        params[["results_dir"]],
        paste0(
            paste(
                "gsea",
                "clusterprofiler",
                "padj",
                alphaThreshold,
                "lfc",
                lfcThreshold,
                sep = "_"
            ),
            ".csv.gz"
        )
    )
)
print(gseaGoTbl)
```

# KEGG enrichment analysis

[Entrez][] identifiers are required for [Kyoto Encyclopedia of Genes and
Genomes (KEGG)][KEGG] analysis. Here we are defining 1:1 mappings of the
[Ensembl][] gene identifiers to [Entrez][] identifiers. For genes that map to
multiple [Entrez][] identifiers, we are using the oldest identifier to define
the 1:1 mapping.

Now let's get the background of unique Entrez identifiers.

```{r all-entrez, eval=evalEntrez}
ensembl2entrez <- Ensembl2Entrez(object = dds, format = "1:1")
allEntrez <- ensembl2entrez[["entrezId"]]
names(allEntrez) <- ensembl2entrez[["ensemblId"]]
str(allEntrez)
```

Let's obtain a vector of significant genes that map to Entrez.

```{r sig-entrez, eval=evalEntrez}
idx <- na.omit(match(x = sigGenes, table = ensembl2entrez[["ensemblId"]]))
sigEntrez <- ensembl2entrez[["entrezId"]][idx]
names(sigEntrez) <- ensembl2entrez[["ensemblId"]][idx]
str(sigEntrez)
```

Note that this vector can contain duplicate Entrez IDs, which we will resolve by significance (adjusted P value).

```{r entrez-lfc-vec, eval=evalEntrez}
entrezLfcVec <- sigRes[names(sigEntrez), "log2FoldChange", drop = TRUE]
names(entrezLfcVec) <- unname(sigEntrez)
## Sort from upregulated to downregulated.
entrezLfcVec <- sort(entrezLfcVec, decreasing = TRUE)
str(entrezLfcVec)
```

```{r enrich-kegg}
kegg <- clusterProfiler::enrichKEGG(
    gene = as.character(sigEntrez),
    organism = keggCode,
    keyType = "ncbi-geneid",
    universe = as.character(allEntrez),
    qvalueCutoff = 0.05
)
keggTbl <-
    kegg |>
    slot("result") |>
    as.data.frame() |>
    camelCase()
saveData(kegg, keggTbl, dir = params[["data_dir"]])
export(
    object = keggTbl,
    file = file.path(
        params[["results_dir"]],
        paste0(
            paste(
                "kegg",
                "clusterprofiler",
                "padj",
                alphaThreshold,
                "lfc",
                lfcThreshold,
                sep = "_"
            ),
            ".csv.gz"
        )
    )
)
print(keggTbl)
```

## KEGG GSEA analysis

[GSEA][] analysis is performed with the [clusterProfiler][] tool using KEGG
gene sets and using the log2 fold changes as input. By using the log2 fold
changes as the input, we are identifying pathways with genes that exhibit
coordinated fold changes that are larger than might be expected by chance. The
significant pathways can be visualized using the log2 fold changes with the
Pathview tool.

Gene set enrichment analysis tools use ranked lists of genes (here ranked by
log2FC) without using a threshold. This allows the tools to use more
information to identify enriched biological processes. The [introduction to
gene set enrichment analysis][] goes into more detail about some of the
advantages of this approach. By using the log2 fold changes as the input, we
are identifying pathways with genes that exhibit coordinated fold changes that
are larger than might be expected by chance. The significant pathways can be
visualized using the log2 fold changes with the [pathview][] tool.

The significantly dysregulated pathways (q-value (FDR) less than `0.05`) are
displayed below in the pathway images, which show the degree of dysregulation
of the genes (the minus direction (green) is down-regulated, while the positive
direction (red) is up-regulated).

When performing [GSEA][] analysis it may be useful to adjust the `minGSSize`
and/or `maxGSSize` parameter based on the pathways you would like to search for
significance. If you are interested in smaller pathways (e.g. a gene set size
of 24 genes), then you would want to adjust the `minGSSize` to less than 24. If
you are only interested in larger pathways, then you would want to adjust the
`GSSize` to a larger number. The fewer pathways tested, the less we need to
correct for multiple test correction, so by adjusting the `minGSSize` and
`maxGSSize` parameters you can test fewer pathways and limit testing to the
pathways of interest.

```{r kegg-gsea, eval=evalEntrez}
gseaKegg <- clusterProfiler::gseKEGG(
    geneList = entrezLfcVec,
    organism = keggCode,
    keyType = "ncbi-geneid"
)
gseaKeggTbl <-
    gseaKegg |>
    slot("result") |>
    as.data.frame() |>
    camelCase()
saveData(gseaKegg, dir = params[["data_dir"]])
export(
    object = gseaKeggTbl,
    file = file.path(
        params[["results_dir"]],
        paste0(
            paste(
                "gsea",
                "kegg",
                "clusterprofiler",
                "padj",
                alphaThreshold,
                "lfc",
                lfcThreshold,
                sep = "_"
            ),
            ".csv.gz"
        )
    )
)
print(gseaKeggTbl)
```

## Pathview {.tabset}

```{r pathview-kegg-plots, message=FALSE, results="asis", eval=evalPathview}
## NOTE This requires Entrez identifiers.
## > help(topic = "pathview", package = "pathview")
##
## There is currently no way to set the output path of the pathview PNG files,
## so we're changing the working directory temporarily using the withr package.
##
## Also, We're using `tryCatch()` here to return to the user any pathways that
## didn't output graphics correctly.
pathways <- gseaKeggTbl[["id"]]
if (hasLength(pathways)) {
    ## `gene.data` vector must be numeric with Entrez identifiers as names.
    ## Using `withTimeout` here to harden against very large pathway sets
    ## (e.g.`"mmu01100"` for F1000 paper example) that can fail to plot.
    ## dplyr may need to be unloaded for pathview to work.
    ## > suppressWarnings(detach("package:dplyr", unload = TRUE, force = TRUE))
    withr::with_dir(
        new = keggPlotsDir,
        code = {
            invisible(lapply(
                X = pathways,
                FUN = function(pathway) {
                    tryCatch(
                        expr = R.utils::withTimeout(
                            expr = {
                                pathview::pathview(
                                    gene.data = entrezLfcVec,
                                    pathway.id = pathway,
                                    gene.idtype = "entrez",
                                    gene.annotpkg = NULL,
                                    species = keggCode,
                                    ## Use bidirectional coloring.
                                    limit = list(gene = 2L)
                                )
                            },
                            timeout = 60L
                        ),
                        error = function(e) {
                            warning(sprintf("%s failed to plot.", pathway))
                        }
                    )
                }
            ))
        }
    )
    figures <- list.files(
        path = keggPlotsDir,
        pattern = "pathview",
        full.names = TRUE
    )
    invisible(lapply(
        X = figures,
        FUN = function(figure) {
            markdownHeader(basename(figure), level = 3L, asis = TRUE)
            cat(paste0("<img src=\"", figure, "\">\n"))
        }
    ))
}
```

```{r footer, child="_footer.Rmd"}
```

```{r links, child="_links.Rmd"}
```

[introduction to gene set enrichment analysis]: http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896
